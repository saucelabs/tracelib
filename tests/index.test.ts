import Tracelib from '../src/index'
import JANK_TRACE_LOG from './__fixtures__/jankTraceLog.json'

let trace: Tracelib
beforeAll(() => {
    trace = new Tracelib(JANK_TRACE_LOG)
})

test('should contain traceLog', () => {
    const sampleTrace = new Tracelib({ foo: 'bar' })
    expect(sampleTrace.tracelog).toEqual({ foo: 'bar' })
})

test('should get FPS', () => {
    const result = trace.getFPS()
    expect(result).toMatchSnapshot()
})

describe('getSummary', () => {
    it('should get summary data', () => {
        const result = trace.getSummary()
        expect(result).toMatchSnapshot()
    })

    it('should not throw error on second call of getSummary', () => {
        const result = trace.getSummary()
        expect(result).toMatchSnapshot()
    })

    it('should get summary data between passed range', () => {
        const result = trace.getSummary(289960055.634, 289960729.717)
        expect(result).toMatchSnapshot()
    })
})

describe('getWarningCounts', () => {
    it('should get warning counts', () => {
        const result = trace.getWarningCounts()
        expect(result).toMatchSnapshot()
    })
})

test('should get memory counters', () => {
    const result = trace.getMemoryCounters()
    expect(result).toMatchSnapshot()
})

describe('mainTrackEvents', () => {
    it('should get events', () => {
        const result = trace.getMainTrackEvents()
        expect(result.length).toEqual(56244)
    })

    it('should throws error if main track is missing', () => {
        /**
         * use tracelog with CrRenderer thread name metadata missing
         */
        const borkedTrace = JANK_TRACE_LOG.slice(0, -1)

        const tracelib = new Tracelib(borkedTrace)
        const result = tracelib.getMainTrackEvents()
        expect(result.length).toEqual(56244)
    })
})

describe('getDetailStats', () => {
    it('should get detail stats', () => {
        const result = trace.getDetailStats()
        expect(Object.keys(result)).toMatchSnapshot()
    })

    it('should not throw error on second call of getDetailStats', () => {
        const result = trace.getDetailStats()
        expect(Object.keys(result)).toMatchSnapshot()
    })

    it('should get summary data between passed range', () => {
        const result = trace.getDetailStats(289960055.634, 289960729.717)
        expect(Object.keys(result)).toMatchSnapshot()
    })
})
