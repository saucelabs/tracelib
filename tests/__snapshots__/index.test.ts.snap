// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDetailStats should get detail stats 1`] = `
Array [
  "rendering",
  "painting",
  "other",
  "scripting",
  "range",
]
`;

exports[`getDetailStats should get summary data between passed range 1`] = `
Array [
  "rendering",
  "painting",
  "other",
  "scripting",
  "range",
]
`;

exports[`getDetailStats should not throw error on second call of getDetailStats 1`] = `
Array [
  "rendering",
  "painting",
  "other",
  "scripting",
  "range",
]
`;

exports[`getSummary should get summary data 1`] = `
Object {
  "endTime": 289961229.717,
  "idle": 52.38300037384033,
  "other": 9.896000564098358,
  "painting": 69.94999980926514,
  "rendering": 847.373997092247,
  "scripting": 394.4800021648407,
  "startTime": 289959855.634,
}
`;

exports[`getSummary should get summary data between passed range 1`] = `
Object {
  "endTime": 289960729.717,
  "idle": 0.6339998841285706,
  "other": 4.653000295162201,
  "painting": 34.8999999165535,
  "rendering": 425.89399832487106,
  "scripting": 208.0020015835762,
  "startTime": 289960055.634,
}
`;

exports[`getSummary should not throw error on second call of getSummary 1`] = `
Object {
  "endTime": 289961229.717,
  "idle": 52.38300037384033,
  "other": 9.896000564098358,
  "painting": 69.94999980926514,
  "rendering": 847.373997092247,
  "scripting": 394.4800021648407,
  "startTime": 289959855.634,
}
`;

exports[`getWarningCounts should get warning counts 1`] = `
Object {
  "ForcedLayout": 4683,
  "ForcedStyle": 4684,
  "LongRecurringHandler": 13,
}
`;

exports[`should get FPS 1`] = `
Array [
  182.2821727559685,
  10.307790628308753,
  11.092131244032895,
  11.014792866762287,
  11.00897231503525,
  10.794939328106791,
  10.929081197725838,
  10.815838712204958,
  10.556652274643293,
  10.47987340271033,
  10.926095888726774,
  10.486577179634944,
  10.897876006628481,
  10.839990888916617,
]
`;

exports[`should get memory counters 1`] = `
Object {
  "documents": Object {
    "times": Array [
      289959946.271,
    ],
    "values": Array [
      5,
    ],
  },
  "gpuMemoryUsedKB": Object {
    "times": Array [],
    "values": Array [],
  },
  "jsEventListeners": Object {
    "times": Array [
      289959946.271,
    ],
    "values": Array [
      38,
    ],
  },
  "jsHeapSizeUsed": Object {
    "times": Array [
      289959946.271,
      289959983.488,
      289960044.53,
      289960134.836,
      289960210.797,
      289960225.218,
      289960316.444,
      289960408.738,
      289960459.387,
      289960499.987,
      289960592.295,
      289960687.485,
      289960694.396,
      289960713.344,
      289960782.934,
      289960874.227,
      289960965.767,
      289960969.893,
      289961061.729,
      289961153.41,
      289961219.984,
    ],
    "values": Array [
      22549008,
      20896736,
      21471160,
      22241560,
      20976176,
      21113456,
      21899112,
      22662104,
      21028744,
      21407840,
      22185296,
      22947872,
      22949192,
      21045640,
      21666816,
      22435800,
      21096152,
      21133192,
      21922608,
      22685512,
      21179632,
    ],
  },
  "nodes": Object {
    "times": Array [
      289959946.271,
    ],
    "values": Array [
      837,
    ],
  },
}
`;
